### PR TITLE
Add geo fields from ECS into client and server attributes

### PR DESCRIPTION
**Pull request by kaiyan-sheng**
Fixes https://github.com/open-telemetry/opentelemetry-specification/issues/2834

## Changes

This PR is to add `geo` fields from Elastic Common Schema(ECS) into otel client and server attributes. These fields can be used to carry data about a specific location such as city name, country name, longitude, latitude, postal code, and ect.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

File: docs/general/attributes.md
Comment (dickwall) [inline docs/general/attributes.md:418]: Would recommend specifying WGS84 standard here to remove any ambiguity, e.g.
```suggestion
| `geo.location.lon` | double | Longitude of the geo location in WGS84. | `-73.61483` | Recommended |
| `geo.location.lat` | double | Latitude of the geo location in WGS84. | `45.505918` | Recommended |
```

**Comment by trisch-me** [comment]
I think it's better to change this PR to add geo to the registry

File: docs/general/attributes.md
Comment (kaiyan-sheng) [reply docs/general/attributes.md]: Done! Thanks!

File: model/registry/client.yaml
Comment (AlexanderWert) [inline model/registry/client.yaml:79]:
```suggestion
          Longitude of the geo location in [WGS84](https://en.wikipedia.org/wiki/World_Geodetic_System#WGS84).
```

File: model/registry/client.yaml
Comment (AlexanderWert) [inline model/registry/client.yaml:84]:
```suggestion
          Latitude of the geo location in [WGS84](https://en.wikipedia.org/wiki/World_Geodetic_System#WGS84).
```

File: model/registry/server.yaml
Comment (AlexanderWert) [inline model/registry/server.yaml:79]:
```suggestion
          Longitude of the geo location in [WGS84](https://en.wikipedia.org/wiki/World_Geodetic_System#WGS84).
```

File: model/registry/server.yaml
Comment (AlexanderWert) [inline model/registry/server.yaml:84]:
```suggestion
          Latitude of the geo location in [WGS84](https://en.wikipedia.org/wiki/World_Geodetic_System#WGS84).
```

File: model/registry/server.yaml
Comment (AlexanderWert) [inline model/registry/server.yaml:104]:
```suggestion
          Region ISO code ([ISO 3166-2](https://en.wikipedia.org/wiki/ISO_3166-2)).
```

File: model/registry/server.yaml
Comment (AlexanderWert) [inline model/registry/server.yaml:69]:
```suggestion
          Two-letter ISO Country Code ([ISO 3166-1 alpha2](https://en.wikipedia.org/wiki/ISO_3166-1#Codes)).
```

File: model/registry/client.yaml
Comment (AlexanderWert) [inline model/registry/client.yaml:104]:
```suggestion
          Region ISO code ([ISO 3166-2](https://en.wikipedia.org/wiki/ISO_3166-2)).
```

File: model/registry/client.yaml
Comment (AlexanderWert) [inline model/registry/client.yaml:114]:
```suggestion
          The time zone of the location, such as [IANA time zone name](https://nodatime.org/TimeZones).
```

File: model/registry/client.yaml
Comment (AlexanderWert) [inline model/registry/client.yaml:69]:
```suggestion
          Two-letter ISO Country Code ([ISO 3166-1 alpha2](https://en.wikipedia.org/wiki/ISO_3166-1#Codes)).
```

File: model/registry/server.yaml
Comment (AlexanderWert) [inline model/registry/server.yaml:114]:
```suggestion
          The time zone of the location, such as [IANA time zone name](https://nodatime.org/TimeZones).
```

**Review by trisch-me — APPROVED** [review]
LGTM, but we will need to think about re-usage of the fields, currently we are repeating them a lot

**Comment by trask** [comment]
can we add `geo.yaml` instead?

**Comment by AlexanderWert** [comment]
> can we add `geo.yaml` instead?

@trask In the registry we structure the files by the top-level namespace (in that case server and client). I'd like to keep it consistent.

But, as soon as we have a way of reusing / nesting attribute definitions under other namespaces, we would move all geo.* fields to a dedicated `geo.yaml` and only reuse in other yaml files like client and server.

**Comment by kaiyan-sheng** [comment]
> can we add `geo.yaml` instead?

In addition to what Alex said, I just want to point out in [ECS](https://www.elastic.co/guide/en/ecs/current/ecs-geo.html), `geo` fields are not expected to be used directly at the root of the events.

**Comment by trask** [comment]
> as soon as we have a way of reusing / nesting attribute definitions under other namespaces

can we use adding the geo fields as motivation to move this forward instead of adding the duplication? or is the proposal that we would still have duplication even after reuse / nesting is introduced?

**Comment by kaiyan-sheng** [comment]
> can we use adding the geo fields as motivation to move this forward instead of adding the duplication? or is the proposal that we would still have duplication even after reuse / nesting is introduced?

We would like to use the geo fields as motivation to move forward with reusing / nesting attribute definitions. That's why we want to add geo fields into 2 attributes (client and server) in this PR to present the need for reusing definitions. 

Do you prefer to have geo fields added into for example only client here, and then enable the reusing attribute definition feature, and then add geo fields for server?

**Comment by trask** [comment]
I was thinking that by introducing reuse / nesting first, we might avoid some churn, but i'm not sure what the plan is for reuse / nesting so definitely could be mistaken

**Comment by kaiyan-sheng** [comment]
@trask We thought about this too but since we don't know when the nested fields will be implemented, we would like to add the geo fields first and then clean up later.

**Comment by trask** [comment]
> we don't know when the nested fields will be implemented

ya, this is a bit of my worry, so trying to leverage the motivation for adding geo fields into adding support for nested fields in general 😄

but really, I don't mean to hold this up, I'm happy to go along with whichever approach @open-telemetry/specs-semconv-maintainers want to go here

**Comment by joaopgrassi** [comment]
I wonder, could we add `geo.yaml` and then still have `geoclient.yaml` and `geoserver.yaml` and those two simple `ref` the attributes from `geo.yaml` but under the prefix `server`/`client` in each file?

**Comment by AlexanderWert** [comment]
> I wonder, could we add `geo.yaml` and then still have `geoclient.yaml` and `geoserver.yaml` and those two simply `ref` the attributes from `geo.yaml` but under the prefix `server`/`client` in each file?

That exactly would be the idea with reuse under new namespace. But, IIUC, the following is not possible with the syntax / tooling yet:

> and those two simply `ref` the attributes from `geo.yaml` but under the prefix `server`/`client` in each file

File: docs/attributes-registry/client.md
Comment (scheler) [inline docs/attributes-registry/client.md:18]: Do we expect any client.geo and server.geo attributes to be present simultaneously?  I thought the purpose of a global attribute registry is to represent attributes that mean the same wherever they are used, and that suits well for the geo.* attributes. That is, geo.city_name means the same whether it's used in the context of client or server or any other that's described at the end at https://www.elastic.co/guide/en/ecs/current/ecs-geo.html

Even if the client.geo.x and server.geo.x attributes are required in the same object, I think this problem is unsolved currently - for eg., one cannot have two `url.full` attributes in the same object if I have to represent the referring url or the url being redirected to in the case of 302.

I think we should omit the `client` and `server` prefix.

**Review by scheler — CHANGES_REQUESTED** [review]

File: docs/attributes-registry/client.md
Comment (AlexanderWert) [reply docs/attributes-registry/client.md:18]: > Even if the client.geo.x and server.geo.x attributes are required in the same object, I think this problem is unsolved currently - for eg., one cannot have two url.full attributes in the same object if I have to represent the referring url or the url being redirected to in the case of 302.

@scheler  This is exactly what we are trying to solve with https://github.com/open-telemetry/semantic-conventions/issues/339. For the geo example that means:
- we will define `geo.*` fields only once in the registry
- and then, `geo.*` fields will be (explicitly) reusable under different namespaces (while keeping their original definition just applied into the context of the target namespace). This will result in `client.geo.*` and `server.geo.*` anyways. 
- this will also solve the problem of having the same fields on a single signal with different context (as they would have different namespaces)

> I think we should omit the client and server prefix.

- We are just defining it here explicitly (for now) because we are still working on the [reuse concept](https://github.com/open-telemetry/semantic-conventions/issues/339) for semantic conventions and don't want to block Geo fields because of that. 
- If you mean that we should not have a prefix for geo at all (even with the reuse concept), then I don't fully get the reasoning, yet. As described above it solves a concrete problem we have today, that you also mentioned yourself with being able to have the same group of attributes twice on a signal with different context. 
- If we would omit the `client`, `server` prefixes for geo, we would break ECS with no reason (no actual conflict with semconv or attribute naming guidelines), which has been explicitly stated as something to avoid whenever possible [in the original OTEP](https://github.com/open-telemetry/oteps/pull/222).

Comment (scheler) [reply docs/attributes-registry/client.md:18]: @AlexanderWert thanks for the clarification. One issue though is that one cannot use `geo.city_name` itself. Is that intended here?

Comment (kaiyan-sheng) [reply docs/attributes-registry/client.md:18]: @scheler Yes it's intended. In [Elastic common schema](https://www.elastic.co/guide/en/ecs/current/ecs-geo.html), `geo` fields are not expected to be used directly at the root of the events.

**Comment by joaopgrassi** [comment]
Hi @kaiyan-sheng !

We changed how the CHANGELOG.md is managed. Please take a look at https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#adding-a-changelog-entry to see what needs to be done. Sorry for the disruption.

**Pull request closed without merging**
